flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractFlinkResourceReconciler.java

```diff
@@ -116,9 +119,23 @@ public final void reconcile(CR cr, Context<?> ctx) throws Exception {
         if (reconciliationStatus.isBeforeFirstDeployment()) {
             LOG.info("Deploying for the first time");
 
+            if (spec.getJob() != null) {
+                var initialUpgradeMode = UpgradeMode.STATELESS;
+                var initialSp = spec.getJob().getInitialSavepointPath();
+
+                if (initialSp != null) {
+                    status.getJobStatus()
+                            .getSavepointInfo()
+                            .setLastSavepoint(
+                                    Savepoint.of(initialSp, SavepointTriggerType.UNKNOWN));
+                    initialUpgradeMode = UpgradeMode.SAVEPOINT;
+                }
+
+                spec.getJob().setUpgradeMode(initialUpgradeMode);
```

Comment:
@gyula-fora this wasn't resolved. I would expect the "reconciled" (and stable) spec to match the spec that was supplied. AFAIK, comparing desired spec with reconciled/stable spec is currently the only way to determine if all changes have been applied. If the specs don't match, how can that be achieved?

Reply:
We already store modified versions of the spec as the operator reconciles the resource. We had similar logic already during suspend/ upgrade/savepoints etc.

this is not a problem because we do not use a naive equality check to test whether user requested changes have been applied. But instead we determine the diff and ignore changes that do not result in an upgrade but only affect reconciliation (such as upgrade mode and operator configs)

Since this is already done and works without any issues and is actually a very clean solution I would suggest we not change it.

Reply:
Also I would note that this only happens if what the user requested in the spec cannot be applied directly but we apply an allowed but different upgrade mode instead . It would be strange if this was not reflected in the status and would make it difficult to build some feature that require knowing what actually was done during last upgrade

Reply:
Makes sense. Agreed that it must be possible to find the actually used upgrade mode. And since upgrade mode change does not cause reconciliation there is no need to consider it when comparing the specs from user side either. That a new mode was applied is evident by its presence in the latest spec.